pkg/util/fedinformer/typedmanager/single-cluster-manager.go

```diff
@@ -204,14 +204,21 @@ func (s *singleClusterInformerManagerImpl) Lister(resource schema.GroupVersionRe
 	}
 	s.lock.Unlock()
 
-	s.lock.RLock()
```

Comment:
I guess the modification of @helen-frank is intended to release the read lock immediately after the current code block is processed. 

If defer is used, it will wait for the lister to be created.

Reply:
> Avoid situations where RUnlock cannot be performed in the presence of an error.

Nice finding!
Yes, when there is an error, here will skip the unlock. 
@helen-frank Just out of curious, how do you find it?

Reply:
HaHa, there was a memory leak in a project that referenced part of karmada's code, and while troubleshooting I found this piece of code and changed it (the failure didn't happen because of this, if it did, it would have been a bugfix)

Reply:
Thank you @helen-frank very much. Appreciate!

Reply:
@whitewindmills Please take another look.
This approach and followings look good to me.

use defer: Given the following `xxx.Lister()` operations are pretty quick, so I don't think there is a problem if expanding the lock scope.
```diff
@@ -205,13 +205,13 @@ func (s *singleClusterInformerManagerImpl) Lister(resource schema.GroupVersionRe
        s.lock.Unlock()

        s.lock.RLock()
+       defer s.lock.RUnlock()
        if resourceTransformFunc, ok := s.transformFuncs[resource]; ok && !s.isInformerStarted(resource) {
                err = resourceInformer.Informer().SetTransform(resourceTransformFunc)
                if err != nil {
                        return nil, err
                }
        }
-       s.lock.RUnlock()

        if resource == nodeGVR {
                return s.informerFactory.Core().V1().Nodes().Lister(), nil
```

no defer: release lock where there is an error.
```diff
@@ -208,6 +208,7 @@ func (s *singleClusterInformerManagerImpl) Lister(resource schema.GroupVersionRe
        if resourceTransformFunc, ok := s.transformFuncs[resource]; ok && !s.isInformerStarted(resource) {
                err = resourceInformer.Informer().SetTransform(resourceTransformFunc)
                if err != nil {
+                       s.lock.RUnlock()
                        return nil, err
                }
        }
```

Reply:
@RainbowMango Same applies to me.

Reply:
got it @RainbowMango